test/modules/TestResultCoordinator/CountingReportGenerator.cs

```diff
@@ -0,0 +1,207 @@
+        readonly int batchSize;
+
+        ulong totalExpectCount = 0;
+        ulong totalMatchCount = 0;
+        ulong totalDuplicateResultCount = 0;
+        List<TestOperationResult> unmatchedResults;
+
+        public CountingReportGenerator(
+            string trackingId,
+            string expectedSource,
+            ISequentialStore<TestOperationResult> expectedStore,
+            string actualSource,
+            ISequentialStore<TestOperationResult> actualStore,
+            string resultType,
+            ITestResultComparer<TestOperationResult> testResultComparer,
+            int batchSize = 500)
+        {
+            this.trackingId = Preconditions.CheckNonWhiteSpace(trackingId, nameof(trackingId));
+            this.expectedSource = Preconditions.CheckNonWhiteSpace(expectedSource, nameof(expectedSource));
+            this.expectedStore = Preconditions.CheckNotNull(expectedStore, nameof(expectedStore));
+            this.actualSource = Preconditions.CheckNonWhiteSpace(actualSource, nameof(actualSource));
+            this.actualStore = Preconditions.CheckNotNull(actualStore, nameof(actualStore));
+            this.resultType = Preconditions.CheckNonWhiteSpace(resultType, nameof(resultType));
+            this.testResultComparer = Preconditions.CheckNotNull(testResultComparer, nameof(testResultComparer));
+            this.batchSize = batchSize;
+        }
+
+        internal string ActualSource => this.actualSource;
+
+        internal ISequentialStore<TestOperationResult> ActualStore => this.actualStore;
+
+        internal string ExpectedSource => this.expectedSource;
+
+        internal ISequentialStore<TestOperationResult> ExpectedStore => this.expectedStore;
+
+        internal string ResultType => this.resultType;
+
+        internal ITestResultComparer<TestOperationResult> TestResultComparer => this.testResultComparer;
+
+        internal int BatchSize => this.batchSize;
+
+        /// <summary>
+        /// Compare 2 data stores and counting expect, match, and duplicate results; and return a counting report.
+        /// It will remove consecutive duplicate results when loading from actual store.
+        /// It will log fail if actual store has more results than expect store.
+        /// </summary>
+        /// <returns></returns>
+        public async Task<ITestResultReport> CreateReportAsync()
+        {
+            Logger.LogInformation($"Start to generate report by {nameof(CountingReportGenerator)} for Sources [{this.expectedSource}] and [{this.actualSource}]");
+
+            long lastLoadedKeyFromExpectStore = -1;
+            long lastLoadedKeyFromActualStore = -1;
+            TestOperationResult lastLoadedResult = default(TestOperationResult);
+
+            var expectQueue = new Queue<TestOperationResult>();
+            var actualQueue = new Queue<TestOperationResult>();
+            totalExpectCount = 0;
+            totalMatchCount = 0;
+            totalDuplicateResultCount = 0;
```

Comment:
>totalDuplicateResultCount [](start = 12, length = 25)

Where  is this incremented?

Reply:
it looks like totalExpectCount, totalMatchCount and totalDuplicateResultCount should be local variables. In that case LoadBatchIntoQueueAsync could also return the number of duplicates for the current load

---
In reply to: [356809881](https://github.com/Azure/iotedge/pull/2103#discussion_r356809881) [](ancestors = 356809881)